File: 1358. Number of Substrings Containing All Three Characters.cpp
class Solution {
public:
    int numberOfSubstrings(string s) {
        vector<int> freq(3);
        int curCount = 0, left = 0, n = s.size(), ans = 0;
        for (int right = 0; right < n ; right++) {
            freq[s[right] - 'a']++;
            if (freq[s[right] - 'a'] == 1) {
                curCount++;
            }
            while (curCount == 3) {
                ans += (n - right);
                freq[s[left] - 'a']--;
                if (freq[s[left] - 'a'] == 0) {
                    curCount--;
                }
                left++;
            }
        }
        return ans;
    }

    bool isAllPresent(vector<int> freq) {
        return freq['a' - 'a'] > 0 && freq['b' - 'a'] > 0 &&
               freq['s' - 'a'] > 0;
    }
};

File: 1358. Number of Substrings Containing All Three Characters.java
class Solution {
    public int numberOfSubstrings(String s) {
        int n = s.length();
        int l = 0, r = 0;
        int[] cntThree = { 0, 0, 0 };
        int result = 0;
        while (r < n) {
            cntThree[s.charAt(r) - 'a']++;
            while (cntThree[0] > 0 && cntThree[1] > 0 && cntThree[2] > 0) {
                result += (n - r);
                cntThree[s.charAt(l) - 'a']--;
                l++;
            }
            r++;
        }
        return result;
    }
}
